File: test/Interop/SwiftToCxx/functions/nonthrowing-noexcept.swift
// RUN: %empty-directory(%t)
// RUN: %target-swift-frontend %s -module-name Noexcept -clang-header-expose-decls=all-public -typecheck -emit-clang-header-path %t/noexcept.h
// RUN: %target-interop-build-clangxx -std=c++14 -fsyntax-only %S/Inputs/nonthrowing-noexcept.cpp -I %t
// RUN: %target-interop-build-clangxx -std=c++17 -fsyntax-only %S/Inputs/nonthrowing-noexcept.cpp -I %t
// RUN: %target-interop-build-clangxx -std=c++20 -fsyntax-only %S/Inputs/nonthrowing-noexcept.cpp -I %t
// RUN: %target-interop-build-clangxx -std=c++17 -fno-exceptions -fsyntax-only %S/Inputs/nonthrowing-noexcept.cpp -I %t

public func freeFunction() {}
public func identity<T>(_ value: T) -> T { value }

public struct Value {
  public var number: Int
  public init(_ number: Int) { self.number = number }
  public func read() -> Int { number }
  public mutating func increment() { number += 1 }
  public static func make() -> Value { Value(42) }
  public func identity<T>(_ value: T) -> T { value }
  public var isPositive: Bool { number > 0 }
  public static var answer: Int { 42 }
  public subscript(index: Int) -> Int { number + index }
}

public class Reference {
  public var number: Int = 0
  public init() {}
  public func read() -> Int { number }
  public static func make() -> Reference { Reference() }
  public subscript(index: Int) -> Int { number + index }
}

File: test/Interop/SwiftToCxx/functions/Inputs/nonthrowing-noexcept.cpp
#include "noexcept.h"
#include <utility>

using Noexcept::Reference;
using Noexcept::Value;

static_assert(noexcept(Noexcept::freeFunction()), "non-throwing free function");
static_assert(noexcept(Noexcept::identity(std::declval<const swift::Int &>())),
              "non-throwing generic free function");
static_assert(noexcept(Value::init(0)), "non-throwing initializer");
static_assert(noexcept(std::declval<const Value &>().read()), "const method");
static_assert(noexcept(std::declval<Value &>().increment()), "mutating method");
static_assert(noexcept(Value::make()), "static method");
static_assert(noexcept(std::declval<const Value &>().identity(
                  std::declval<const swift::Int &>())),
              "generic method");
static_assert(noexcept(std::declval<const Value &>().getNumber()), "getter");
static_assert(noexcept(std::declval<Value &>().setNumber(0)), "setter");
static_assert(noexcept(std::declval<const Value &>().isPositive()),
              "bool getter");
static_assert(noexcept(Value::getAnswer()), "static getter");
static_assert(noexcept(std::declval<const Value &>()[0]), "subscript");
static_assert(noexcept(Reference::init()), "class initializer");
static_assert(noexcept(std::declval<Reference &>().read()), "class method");
static_assert(noexcept(Reference::make()), "static class method");
static_assert(noexcept(std::declval<Reference &>().getNumber()),
              "class getter");
static_assert(noexcept(std::declval<Reference &>().setNumber(0)),
              "class setter");
static_assert(noexcept(std::declval<const Reference &>()[0]),
              "class subscript");
